README.md
<p align="center">
  <img alt="tokenpress" src="https://image.flaticon.com/icons/svg/929/929429.svg" width="120">
</p>

<h3 align="center">
  tokenpress
</h3>

<p align="center">
  A JWT utility belt for JavaScript applications
</p>

<p align="center">
  <a href="https://github.com/justinsisley/tokenpress/blob/master/CHANGELOG.md"><img src="https://img.shields.io/github/release/justinsisley/tokenpress.svg?style=for-the-badge" alt="GitHub release" /></a> <a href="https://circleci.com/gh/justinsisley/tokenpress"><img src="https://img.shields.io/circleci/project/github/justinsisley/tokenpress.svg?style=for-the-badge" alt="CircleCI" /></a> <a href="https://github.com/justinsisley/tokenpress/blob/master/LICENSE"><img src="https://img.shields.io/github/license/justinsisley/tokenpress.svg?style=for-the-badge" alt="license" /></a>
</p>

---

# Table of Contents

- [Features](#features)
- [Documentation](#documentation)
  - [Installation](#installation)
  - [Node.js](#nodejs)
  - [Browser](#browser)
  - [Contributing](#contributing)
    - [Linting](#linting)
    - [Testing](#testing)
- [Releases](https://github.com/justinsisley/tokenpress/blob/master/CHANGELOG.md)
- [Credits](#credits)

# Features

- __Convenient, universal utilities for handling JWTs__
- __JWTs generated by [node-jsonwebtoken](https://github.com/auth0/node-jsonwebtoken)__
- __Runs on Node.js v8+__

# Documentation

## Installation

```bash
npm install tokenpress
```

## Node.js

Configure tokenpress before using it:

```javascript
const tokenpress = require('tokenpress');

tokenpress.configure({
  // Required: string or buffer containing the secret for HMAC algorithms
  secret: 'CHANGE_THIS_SECRET',
  // Required: string describing a time span zeit/ms. Eg: 60, "2 days", "10h", "7d"
  expiresIn: '30 days',
  // Optional: Minimum and maximum token lengths for getURLSafeToken utility
  minTokenLength: 30,
  maxTokenLength: 50,
});
```

Sign a token:

```javascript
const tokenpress = require('tokenpress');

const token = tokenpress.jwt.sign({
  username: 'clever_username_ftw',
  role: 'USER',
});
```

Verify a token using JWKS:

```javascript
const tokenpress = require('tokenpress');

tokenpress.configure({
  algorithms: ['RS256'],
  audience: 'my audience',
  issuer: `https://my-app.com/`,
  jwksUri: `https://my-app.com/jwks.json`,
});

const someToken = 'blah.blah.blah';
tokenpress.jwt.verifyWithJWKS(someToken).then((decodedJWT) => {
  console.log(decodedJWT)
});
```

Use tokenpress middleware to require authentication for a route:

```javascript
const tokenpress = require('tokenpress');
const { requireAuth } = tokenpress.middleware;

router.get('/user/account', requireAuth, (req, res) => {
  // req.jwt contains the decoded JWT
  const { username, role } = req.jwt;

  res.json({ username, role });
});
```

> Note: If the authentication check fails, a 401 (unauthorized) response will be sent as JSON. The response will contain an `error` property that will equal either `EXPIRED_TOKEN` or `INVALID_TOKEN`. `INVALID_TOKEN` can be caused by any of the conditions listed in the [jsonwebtoken docs](https://github.com/auth0/node-jsonwebtoken#jsonwebtokenerror).

Generate a random, variable-length, hexadecimal string using the crypto.randomBytes function. The minumum length defaults to 30, and the maximum length defaults to 50.

```javascript
const tokenpress = require('tokenpress');

const randomToken = tokenpress.utils.getURLSafeToken();
```

## Browser

Optionally configure whether to use sessionStorage as opposed to localStorage for storing tokens on the client. By default, localStorage will be used.

```javascript
import tokenpress from 'tokenpress/browser';

tokenpress.configure({
  useSessionStorage: true,
});
```

Optionally configure the key used when saving to localStorage or sessionStorage. Defaults to `token`.

```javascript
import tokenpress from 'tokenpress/browser';

tokenpress.configure({
  storageKey: 'custom-token-name',
});
```

Save a token to localStorage/sessionStorage:

```javascript
import tokenpress from 'tokenpress/browser';

mockFunctionToGetTokenFromServer().then((token) => {
  tokenpress.save(token)
});
```

Retrieve a token from localStorage/sessionStorage:

```javascript
import tokenpress from 'tokenpress/browser';

const token = tokenpress.get();
```

Delete a token from localStorage/sessionStorage:

```javascript
import tokenpress from 'tokenpress/browser';

tokenpress.delete();
```

Determine if a token is expired:

```javascript
import tokenpress from 'tokenpress/browser';

// Will fetch token from localStorage/sessionStorage by default
const isTokenExpired = tokenpress.isExpired();
console.log(isTokenExpired); // true or false

// Or, check a token you've previously retrieved
const token = tokenpress.get();
const isMyOtherTokenExpired = tokenpress.isExpired(token);
console.log(isMyOtherTokenExpired); // true or false
```

# Contributing

## Linting

Run ESLint with `npm run lint`.

## Testing

Run unit tests with `npm test`.

# Credits
<div>Icon made by <a href="http://www.freepik.com" title="Freepik">Freepik</a> from <a href="https://www.flaticon.com/" title="Flaticon">www.flaticon.com</a> is licensed by <a href="http://creativecommons.org/licenses/by/3.0/" title="Creative Commons BY 3.0" target="_blank">CC 3.0 BY</a></div>


CHANGELOG.md
# Changelog

## v2.3.0

- __docs:__ Updated docs for JWKS [7c9c646]
- __feat:__ Added verification using JWKS [f63c246]


## v2.2.0

- __feat:__ Remove bearer prefix in auth header [d1761a7]


## v2.1.1

- __build:__ Updated dependencies and fixed tests [a55ee6c]


## v2.1.0

- __build:__ Updated dependencies [fa43f64]
- __feat:__ Allow usage of sessionStorage as opposed to localStorage in browser [76dfd70]


## v2.0.1

- __docs:__ Updated badge links [c2f65bf]
- __docs:__ Added link to changelog [2aba9c0]
- __docs:__ Cleaned up changelog [ed6f514]


## v2.0.0

- __build:__ Added dutchie [94761de]
- __refactor:__ Refactored [cee3589]

test.js
const browser = require('./lib/browser');
const node = require('./lib/node');

it('should export an object', () => {
  expect(typeof browser).toBe('object');
});

it('should allow the localStorage key to be customized', () => {
  const sampleToken = '12345';

  // Save to the default key
  browser.save(sampleToken);

  // Change the key
  browser.configure({
    localStorageKey: 'test-key',
  });

  // No token exists for the new key
  expect(browser.get()).toBe(null);

  // Save under the new key and verify
  browser.save(sampleToken);
  expect(browser.get()).toBe(sampleToken);

  // Cleanup
  browser.delete();
});

it('should save, get, and delete tokens', () => {
  const sampleToken = '12345';

  // Should not exist yet
  expect(browser.get()).toBe(null);

  // Save it and verify it's there
  browser.save(sampleToken);
  expect(browser.get()).toBe(sampleToken);

  // Delete it and verify it's gone
  browser.delete();
  expect(browser.get()).toBe(null);
});

it('should verify if a token has expired', (done) => {
  // Configure with a 1-second expiration
  node.configure({
    secret: 'test',
    expiresIn: 1,
  });

  // Create a new token
  node.jwt.sign({
    username: 'test_username',
  }).then((token) => {
    // Save it to localStorage
    browser.save(token);

    // Make sure it hasn't expired yet
    expect(browser.isExpired()).toBe(false);
    expect(browser.isExpired(token)).toBe(false);

    setTimeout(() => {
    // It should now be expired
      expect(browser.isExpired()).toBe(true);
      expect(browser.isExpired(token)).toBe(true);

      // Cleanup
      browser.delete();

      done();
    }, 1000);
  });
});


index.js
const jsonwebtoken = require('jsonwebtoken');

let config = {
  // String or buffer containing the secret for HMAC algorithms
  secret: null,
  // String describing a time span zeit/ms. Eg: 60, "2 days", "10h", "7d"
  expiresIn: null,
  // Key used to save the JWT to the browser's localStorage
  localStorageKey: 'token',
};

const tokenpress = {
  node: {
    configure: ({ secret, expiresIn }) => {
      config = Object.assign({}, config, {
        secret,
        expiresIn,
      });
    },

    middleware: {
      requireAuth(req, res, next) {
        if (!config.secret) {
          throw new Error('missing tokenpress secret');
        }

        const failed = () => res.status(401).json({});

        // Allow query parameter auth token, which takes precedence over headers
        const queryToken = req.query.token;
        if (queryToken) {
          // eslint-disable-next-line
          req.headers.authorization = queryToken;
        }

        const token = req.headers.authorization;
        if (!token) {
          failed();
          return;
        }

        jsonwebtoken.verify(token, config.secret, (err, decodedToken) => {
          if (err) {
            failed();
            return;
          }

          // eslint-disable-next-line no-param-reassign
          req.jwt = decodedToken;

          next();
        });
      },
    },

    jwt: {
      sign(payload) {
        if (!config.secret) {
          throw new Error('missing tokenpress secret');
        }

        if (!config.expiresIn) {
          throw new Error('missing tokenpress expiresIn');
        }

        const secret = config.secret;
        const expiresIn = config.expiresIn;

        return new Promise((resolve, reject) => {
          jsonwebtoken.sign(payload, secret, { expiresIn }, (err, token) => {
            if (err) {
              return reject(err);
            }

            return resolve(token);
          });
        });
      },

      verify(token) {
        if (!config.secret) {
          throw new Error('missing tokenpress secret');
        }

        const secret = config.secret;

        return new Promise((resolve, reject) => {
          jsonwebtoken.verify(token, secret, (err, decoded) => {
            if (err) {
              return reject(err);
            }

            return resolve(decoded);
          });
        });
      },
    },
  },

  browser: {
    configure: ({ localStorageKey }) => {
      config = Object.assign({}, config, {
        localStorageKey,
      });
    },

    save(token) {
      window.localStorage.setItem(config.localStorageKey, token);
    },

    get() {
      return window.localStorage.getItem(config.localStorageKey);
    },

    delete() {
      window.localStorage.removeItem(config.localStorageKey);
    },
  },
};

module.exports = tokenpress;


lib/browser.js
let config = {
  storageKey: 'token',
  useSessionStorage: false,
};

let storage = 'localStorage';

const tokenpress = {
  configure: ({
    storageKey,
    useSessionStorage,
  }) => {
    config = { ...config, storageKey };

    if (useSessionStorage) {
      storage = 'sessionStorage';
    }
  },

  save(token) {
    window[storage].setItem(config.storageKey, token);
  },

  get() {
    return window[storage].getItem(config.storageKey);
  },

  delete() {
    window[storage].removeItem(config.storageKey);
  },

  isExpired(token = window[storage].getItem(config.storageKey)) {
    if (!token) { return true; }

    const split = token.split('.');
    if (split.length < 3) { return true; }

    let payload;
    try {
      payload = JSON.parse(window.atob(split[1]));
    } catch (error) { /* no-op */ }

    if (!payload || !payload.exp) { return true; }

    const expiration = payload.exp;

    const expTimestamp = +(`${expiration}`.padEnd(13, '0'));
    const nowTimestamp = Date.now();

    return nowTimestamp > expTimestamp;
  },

  decode(token) {
    if (!token || !token.split) return null;

    const split = token.split('.');
    if (split.length < 3) return null;

    let payload = null;
    try {
      payload = JSON.parse(window.atob(split[1]));
    } catch (error) { /* no-op */ }

    return payload;
  },
};

module.exports = tokenpress;


lib/node.js
const jsonwebtoken = require('jsonwebtoken');
const jwks = require('jwks-rsa');
const crypto = require('crypto');

let config = {
  // String or buffer containing the secret for HMAC algorithms
  secret: null,

  // String describing a time span or seconds. Eg: 60, "2 days", "10h", "7d"
  expiresIn: null,

  // Min and max token lengths for getURLSafeToken utility
  minTokenLength: 30,
  maxTokenLength: 50,
};

const tokenpress = {
  configure: ({
    secret,
    expiresIn,
    jwksUri,
    audience,
    issuer,
    algorithms,
  }) => {
    config = {
      ...config,
      secret,
      expiresIn,
      audience,
      issuer,
      algorithms,
    };

    if (jwksUri) {
      config.jwksClient = jwks({
        cache: true,
        rateLimit: true,
        jwksUri,
      });
    }
  },

  middleware: {
    requireAuth(req, res, next) {
      if (!config.secret) {
        throw new Error('missing tokenpress secret');
      }

      const failed = (error) => res.status(401).json({ error });

      // Allow query parameter auth token, which takes precedence over headers
      const queryToken = req.query.token;
      if (queryToken) {
        // eslint-disable-next-line
        req.headers.authorization = queryToken;
      }

      let token = req.headers.authorization;
      if (!token) {
        failed();
        return;
      }

      // Remove "Bearer: " prefix if it exists
      token = token.replace(/^bearer: /i, '');

      jsonwebtoken.verify(token, config.secret, (err, decodedToken) => {
        if (err) {
          switch (err.name) {
            case 'TokenExpiredError':
              failed('EXPIRED_TOKEN');
              return;

            case 'JsonWebTokenError':
              failed('INVALID_TOKEN');
              return;

            default:
              failed(err.name);
              return;
          }
        }

        // eslint-disable-next-line no-param-reassign
        req.jwt = decodedToken;

        next();
      });
    },
  },

  jwt: {
    sign(payload) {
      if (!config.secret) {
        throw new Error('missing tokenpress secret');
      }

      if (!config.expiresIn) {
        throw new Error('missing tokenpress expiresIn');
      }

      const { secret, expiresIn } = config;

      return new Promise((resolve, reject) => {
        jsonwebtoken.sign(payload, secret, { expiresIn }, (err, token) => {
          if (err) {
            return reject(err);
          }

          return resolve(token);
        });
      });
    },

    verify(token) {
      if (!config.secret) {
        throw new Error('missing tokenpress secret');
      }

      const { secret } = config;

      return new Promise((resolve, reject) => {
        jsonwebtoken.verify(token, secret, (err, decoded) => {
          if (err) {
            return reject(err);
          }

          return resolve(decoded);
        });
      });
    },

    verifyWithJWKS(token) {
      if (!config.jwksClient) {
        throw new Error('missing JWKS client');
      }

      const getSigningKey = (kid) => new Promise((resolve, reject) => {
        config.jwksClient.getSigningKey(kid, (error, key) => {
          if (error) {
            reject(error);
            return;
          }

          const signingKey = key.publicKey || key.rsaPublicKey;
          resolve(signingKey);
        });
      });

      const verifyJWT = (jwt, signingKey) => new Promise((resolve, reject) => {
        jsonwebtoken.verify(
          jwt,
          signingKey,
          {
            algorithms: config.algorithms,
            audience: config.audience,
            issuer: config.issuer,
          },
          (error, decoded) => {
            if (error) {
              reject(error);
              return;
            }

            resolve(decoded);
          },
        );
      });

      return new Promise((resolve, reject) => {
        const decoded = jsonwebtoken.decode(token, { complete: true });

        getSigningKey(decoded.header.kid).then((signingKey) => {
          verifyJWT(token, signingKey).then(resolve).catch(reject);
        }).catch(reject);
      });
    },
  },

  utils: {
    getURLSafeToken() {
      const { maxTokenLength, minTokenLength } = config;

      return new Promise((resolve, reject) => {
        const random = (Math.random() * (maxTokenLength - minTokenLength)) + minTokenLength;
        const length = Math.floor(random);

        crypto.randomBytes(length, (error, buffer) => {
          if (error) {
            return reject(error);
          }

          return resolve(buffer.toString('hex'));
        });
      });
    },
  },
};

module.exports = tokenpress;
